src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c

```diff
@@ -1124,28 +1098,48 @@ static int xocl_init_memory_manager(struct xocl_drm *drm_p)
 	/* Initialize with max and min possible value */
         mm_start_addr = 0xffffFFFFffffFFFF;
         mm_end_addr = 0;
-	drm_p->cma_bank_idx = -1;
 
         /* Initialize all the banks and their sizes */
         /* Currently only fixed sizes are supported */
         for (i = 0; i < topo->m_count; i++) {
                 mem_data = &topo->m_mem_data[i];
                 ddr_bank_size = mem_data->m_size * 1024;
 
+		if (!ddr_bank_size)
+			continue;
+
 		if (XOCL_IS_STREAM(topo, i))
                         continue;
 
-                if (XOCL_IS_PS_KERNEL_MEM(topo, i))
-                        continue;
+		if (XOCL_IS_PS_KERNEL_MEM(topo, i))
+			continue;
 
-                if (!is_mem_region_valid(drm_p, mem_data))
-                        continue;
+		if (!is_mem_region_valid(drm_p, mem_data))
+			continue;
 
-                /* Update the start and end address for the memory manager */
-                if (mem_data->m_base_address < mm_start_addr)
-                        mm_start_addr = mem_data->m_base_address;
-                if ((mem_data->m_base_address + ddr_bank_size) > mm_end_addr)
-                        mm_end_addr = mem_data->m_base_address + ddr_bank_size;
+		/* Update the start and end address for the memory manager */
+		if (mem_data->m_base_address < mm_start_addr)
+			mm_start_addr = mem_data->m_base_address;
+		if ((mem_data->m_base_address + ddr_bank_size) > mm_end_addr)
+			mm_end_addr = mem_data->m_base_address + ddr_bank_size;
+	}
+
+	/* Memory manager initialize should be done only once */
+	if (drm_p->xocl_mm_done && (drm_p->xocl_mm != NULL)) {
+		/* Validate the new memory topology with existing memory manager */
+		if ((drm_p->xocl_mm->start_addr != mm_start_addr) ||
+				(drm_p->xocl_mm->end_addr != mm_end_addr) ||
+				(drm_p->xocl_mm->m_count != topo->m_count)) {
+			xocl_warn(drm_p->ddev->dev, "Memory Topology doesn't "
+				"consistent in xclbins. Re-initializing Memory Manager");
```

Comment:
This is confusing... At line 1127, the comment says that the memory manager should be done only once. But, here you're doing it again? Is this expected? If so, why issue an warning? If not, then is it really OK to re-init it? Claiming that an entity can only be initialized once is big decision to make and has big impact to future code maintenance. Please think through.